src/Compilers/CSharp/Portable/Binder/RefSafetyAnalysis.cs

```diff
@@ -0,0 +1,914 @@
+        {
+            var fieldEqualsValue = (BoundFieldEqualsValue)initializer;
+
+            using var _ = new LocalScope(this, fieldEqualsValue.Locals);
+
+            base.Visit(fieldEqualsValue.Value);
+
+            var field = fieldEqualsValue.Field;
+            bool isByRef = field.RefKind != RefKind.None;
+            if (isByRef || field.Type.IsRefLikeType)
+            {
+                ValidateEscape(fieldEqualsValue.Value, Binder.CallingMethodScope, isByRef: isByRef, _diagnostics);
+            }
+        }
+
+        public override BoundNode? VisitLocalFunctionStatement(BoundLocalFunctionStatement node)
+        {
+            var localFunction = node.Symbol;
+            // https://github.com/dotnet/roslyn/issues/65353: We should not reuse _localEscapeScopes
+            // across nested local functions or lambdas because _localScopeDepth is reset when entering
+            // the function or lambda so the scopes across the methods are unrelated.
+            var analysis = new RefSafetyAnalysis(_compilation, localFunction, localFunction.IsUnsafe, _useUpdatedEscapeRules, _diagnostics, _localEscapeScopes);
+            analysis.Visit(node.BlockBody);
+            analysis.Visit(node.ExpressionBody);
+            return null;
+        }
+
+        public override BoundNode? VisitLambda(BoundLambda node)
+        {
+            var lambda = node.Symbol;
+            // https://github.com/dotnet/roslyn/issues/65353: We should not reuse _localEscapeScopes
+            // across nested local functions or lambdas because _localScopeDepth is reset when entering
+            // the function or lambda so the scopes across the methods are unrelated.
+            var analysis = new RefSafetyAnalysis(_compilation, lambda, _inUnsafeRegion, _useUpdatedEscapeRules, _diagnostics, _localEscapeScopes);
+            analysis.Visit(node.Body);
+            return null;
+        }
+
+        public override BoundNode? VisitConstructorMethodBody(BoundConstructorMethodBody node)
+        {
+            using var _ = new LocalScope(this, node.Locals);
+            return base.VisitConstructorMethodBody(node);
+        }
+
+        public override BoundNode? VisitForStatement(BoundForStatement node)
+        {
+            using var outerLocals = new LocalScope(this, node.OuterLocals);
+            using var innerLocals = new LocalScope(this, node.InnerLocals);
+            return base.VisitForStatement(node);
+        }
+
+        public override BoundNode? VisitUsingStatement(BoundUsingStatement node)
+        {
+            using var _ = new LocalScope(this, node.Locals);
+
+            var placeholders = ArrayBuilder<(BoundValuePlaceholderBase, uint)>.GetInstance();
+            if (node.AwaitOpt is { } awaitableInfo)
+            {
+                uint valEscapeScope = node.ExpressionOpt is { } expr
+                    ? GetValEscape(expr, _localScopeDepth)
+                    : _localScopeDepth;
```

Comment:
>: _localScopeDepth; [](http://example.com/codeflow?start=20&length=19)

Two questions:
1. that `else` clause was not obvious. Should we do `if (node.ExpressionOpt is { } expr) { GetAwaitableInstancePlaceholders(placeholders, awaitableInfo, valEscapeScope: GetValEscape(expr, _localScopeDepth)); }`
2. Are we testing `BoundUsingStatement` with a `DeclarationOpt` instead of an `ExpressionOpt`?

Reply:
Closing this. Once we refactor the Get/Check methods into the Visit logic, this asymmetry will go away. This code will handle both expression and declaration case, to visit the RHS expression.